--- configure
+++ configure
@@ -4403,7 +4403,7 @@


 if test -z "$user_redefine_cc" && \
-   test "$cross_compiling" != yes && \
+   test "$cross_compiling" = no && \
    test "${with_mini_gmp+set}" != set ; then

 if test -z "$GMP_CC$GMP_CFLAGS" ; then
@@ -7369,6 +7369,10 @@
 os2*)
   lt_cv_deplibs_check_method=pass_all
   ;;
+eklat*)
+  lt_cv_deplibs_check_method=pass_all
+  ;;
+
 esac
 
 fi
@@ -11218,6 +11222,9 @@
   linux* | k*bsd*-gnu | gnu*)
     link_all_deplibs=no
     ;;
+  eklat*)
+    link_all_deplibs=no
+    ;;
   esac
 
   ld_shlibs=yes
@@ -12416,6 +12423,9 @@
       hardcode_shlibpath_var=no
       ;;
 
+    eklat*)
+      ;;
+
     *)
       ld_shlibs=no
       ;;
@@ -13500,6 +13510,15 @@
   shlibpath_var=LD_LIBRARY_PATH
   ;;
 
+eklat*)
+  version_type=linux # correct to gnu/linux during the next big refactor
+  library_names_spec='$libname$release$shared_ext$versuffix $libname$release$shared_ext$major $libname$shared_ext'
+  soname_spec='$libname$release$shared_ext$major'
+  shlibpath_var=LD_LIBRARY_PATH
+  hardcode_into_libs=yes
+  hardcode_libdir_flag_spec=''
+  ;;
+
 *)
   dynamic_linker=no
   ;;
