--- grub-core/loader/multiboot_elfxx.c
+++ grub-core/loader/multiboot_elfxx.c
@@ -238,12 +238,8 @@
   if (i == phnum)
     return grub_error (GRUB_ERR_BAD_OS, "entry point isn't in a segment");
 
-#if defined (__i386__) || defined (__x86_64__)
-
-#elif defined (__mips)
+#ifdef __mips
   GRUB_MULTIBOOT (payload_eip) |= 0x80000000;
-#else
-#error Please complete this
 #endif
 
   if (shnum)
--- grub-core/loader/multiboot_mbi2.c
+++ grub-core/loader/multiboot_mbi2.c
@@ -726,14 +726,10 @@
     return err;
 
   ptrorig = get_virtual_current_address (ch);
-#if defined (__i386__) || defined (__x86_64__)
   *target = get_physical_target_address (ch);
-#elif defined (__mips)
-  *target = get_physical_target_address (ch) | 0x80000000;
-#else
-#error Please complete this
+#ifdef __mips
+  *target |= 0x80000000;
 #endif
-
   mbistart = ptrorig;
   COMPILE_TIME_ASSERT ((2 * sizeof (grub_uint32_t))
 		       % sizeof (grub_properly_aligned_t) == 0);
